File: src/main/java/org/goafabric/calleeservice/extensions/TenantContext.java
package org.goafabric.calleeservice.extensions;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;

import java.io.IOException;
import java.util.Base64;
import java.util.Map;

public class TenantContext {
    record TenantContextRecord(String tenantId, String organizationId, String userName) {
        public Map<String, String> toAdapterHeaderMap() {
            return Map.of("X-TenantId", tenantId, "X-OrganizationId", organizationId, "X-Auth-Request-Preferred-Username", userName);
        }
    }

    private static final ThreadLocal<TenantContextRecord> CONTEXT =
            ThreadLocal.withInitial(() -> new TenantContextRecord("0", "0", "anonymous"));

    public static void setContext(HttpServletRequest request) {
        setContext(request.getHeader("X-TenantId"), request.getHeader("X-OrganizationId"),
                request.getHeader("X-Auth-Request-Preferred-Username"), request.getHeader("X-UserInfo"));
    }

    static void setContext(String tenantId, String organizationId, String userName, String userInfo) {
        CONTEXT.set(new TenantContextRecord(
                getValue(tenantId, "0"),
                getValue(organizationId, "0"),
                getValue(getUserNameFromUserInfo(userInfo), getValue(userName, "anonymous"))
        ));
    }

    public static void removeContext() {
        CONTEXT.remove();
    }

    private static String getValue(String value, String defaultValue) {
        return value != null ? value : defaultValue;
    }

    public static String getTenantId() {
        return CONTEXT.get().tenantId();
    }

    public static String getOrganizationId() {
        return CONTEXT.get().organizationId();
    }

    public static String getUserName() {
        return CONTEXT.get().userName();
    }

    public static Map<String, String> getAdapterHeaderMap() {
        return CONTEXT.get().toAdapterHeaderMap();
    }

    public static void setTenantId(String tenant) {
        CONTEXT.set(new TenantContextRecord(tenant, CONTEXT.get().organizationId, CONTEXT.get().userName));
    }

    private static String getUserNameFromUserInfo(String userInfo) {
        try {
            return userInfo != null ? (String) new ObjectMapper().readValue(Base64.getUrlDecoder().decode(userInfo), Map.class).get("preferred_username") : null;
        } catch (IOException e) {
            throw new IllegalStateException(e);
        }
    }
}

File: src/main/kotlin/org/goafabric/calleeservice/extensions/TenantContext.kt
package org.goafabric.calleeservice.extensions

import jakarta.servlet.http.HttpServletRequest
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder

object TenantContext {
    data class TenantContextRecord(val tenantId: String?, val organizationId: String?, val userName: String?) {
        fun toAdapterHeaderMap(): Map<String, String?> {
            return java.util.Map.of(
                "X-TenantId", tenantId,
                "X-OrganizationId", organizationId,
                "X-Auth-Request-Preferred-Username", userName
            )
        }
    }

    private val tenantContext: ThreadLocal<TenantContextRecord> =
        ThreadLocal.withInitial { TenantContextRecord(null, null, null) }

    fun setContext(request: HttpServletRequest) {
        tenantContext.set(
            TenantContextRecord(
                request.getHeader("X-TenantId"),
                request.getHeader("X-OrganizationId"),
                request.getHeader("X-Auth-Request-Preferred-Username")
            )
        )
    }

    fun setContext(tenantContextRecord: TenantContextRecord) {
        tenantContext.set(tenantContextRecord)
    }

    fun removeContext() {
        tenantContext.remove()
    }

    var tenantId: String
        get() = if (tenantContext.get().tenantId != null) tenantContext.get().tenantId!! else "0"
        set(tenant) {
            tenantContext.set(
                TenantContextRecord(
                    tenant,
                    tenantContext.get().organizationId,
                    tenantContext.get().userName
                )
            )
        }

    val organizationId: String
        get() = if (tenantContext.get().organizationId != null) tenantContext.get().organizationId!! else "0"

    val userName: String
        get() = if ((authentication != null) && authentication!!.name != "anonymousUser") authentication!!.name
        else (if (tenantContext.get().userName != null) tenantContext.get().userName else "anonymous")!!

    val adapterHeaderMap: Map<String, String?>
        get() = tenantContext.get().toAdapterHeaderMap()

    private val authentication: Authentication?
        get() = SecurityContextHolder.getContext().authentication
}
